include/smack/smack.h

```diff
@@ -195,8 +461,40 @@ void __SMACK_decls() {
     "  }\n"
     "  $Alloc[p] := true;\n"
     "}");
+#endif
 
 #elif MEMORY_MODEL_REUSE // can reuse previously-allocated and freed addresses
+#ifdef BITVECTOR
+  D("var $Alloc: [ref] bool;");
+  D("var $Size: [ref] size;");
+
+  D("procedure $malloc(n: size) returns (p: ref);\n"
+    "modifies $Alloc, $Size;\n"
+    "ensures $sgt.ref(p, $NULL);\n"
+    "ensures !old($Alloc[p]);\n"
+    "ensures (forall q: ref :: old($Alloc[q]) ==> ($slt.ref($add.ref(p, n), q) || $sgt.ref(p, $add.ref(q, $Size[q]))));\n"
+    "ensures $Alloc[p];\n"
+    "ensures $Size[p] == n;\n"
+    "ensures (forall q: ref :: {$Size[q]} q != p ==> $Size[q] == old($Size[q]));\n"
+    "ensures (forall q: ref :: {$Alloc[q]} q != p ==> $Alloc[q] == old($Alloc[q]));\n"
+    "ensures $sge.ref(n, $NULL) ==> (forall q: ref :: {$base(q)} $sle.ref(p, q) && $slt.ref(q, $add.ref(p, n)) ==> $base(q) == p);");
+
+  D("procedure $free(p: ref);\n"
+    "modifies $Alloc;\n"
+    "ensures !$Alloc[p];\n"
+    "ensures (forall q: ref :: {$Alloc[q]} q != p ==> $Alloc[q] == old($Alloc[q]));");
+
+  D("procedure $alloca(n: size) returns (p: ref);\n"
+    "modifies $Alloc, $Size;\n"
+    "ensures $sgt(p, $NULL);\n"
+    "ensures !old($Alloc[p]);\n"
+    "ensures (forall q: ref :: old($Alloc[q]) ==> ($slt.ref($add.ref(p, n), q) || $sgt.ref(p, $add.ref(q, $Size[q]))));\n"
+    "ensures $Alloc[p];\n"
+    "ensures $Size[p] == n;\n"
+    "ensures (forall q: ref :: {$Size[q]} q != p ==> $Size[q] == old($Size[q]));\n"
+    "ensures (forall q: ref :: {$Alloc[q]} q != p ==> $Alloc[q] == old($Alloc[q]));\n"
+    "ensures $sge(n, $NULL) ==> (forall q: ref :: {$base(q)} $sle.ref(p, q) && $slt.ref(q, $add.ref(p, n)) ==> $base(q) == p);");
+#else
```

Comment:
As noted above, the types `ref` and `size` should replace all uses of `int` for pointer types and size types. The following declarations break this rule.